File: frontend/src/services/questionService.ts
import { gql } from "@apollo/client/core";
import { apolloClient } from "../apollo-client.ts";

export interface Question {
  id: string;
  quizId: string;
  question: string;
  image?: string;
  createdAt: string;
}

export const questionService = {
  getQuestion: async (id: string) => {
    const { data } = await apolloClient.query({
      query: gql`
        query GetQuestion($id: ID!) {
          question(id: $id) {
            id
            quizId
            question
            image
            createdAt
          }
        }
      `,
      variables: { id },
    });
    return data.question;
  },

  getQuestions: async (quizId: string) => {
    const { data } = await apolloClient.query({
      query: gql`
        query GetQuestions($quizId: ID!) {
          questions(quizId: $quizId) {
            id
            quizId
            question
            image
            createdAt
          }
        }
      `,
      variables: { quizId },
    });
    return data.questions;
  },

  createQuestion: async (quizId: string, question: string, image?: string) => {
    const { data } = await apolloClient.mutate({
      mutation: gql`
        mutation CreateQuestion(
          $quizId: ID!
          $question: String!
          $image: String
        ) {
          createQuestion(quizId: $quizId, question: $question, image: $image) {
            id
            quizId
            question
            image
            createdAt
          }
        }
      `,
      variables: { quizId, question, image },
    });
    return data.createQuestion;
  },

  updateQuestion: async (id: string, question?: string, image?: string) => {
    const { data } = await apolloClient.mutate({
      mutation: gql`
        mutation UpdateQuestion($id: ID!, $question: String, $image: String) {
          updateQuestion(id: $id, question: $question, image: $image) {
            id
            quizId
            question
            image
            createdAt
          }
        }
      `,
      variables: { id, question, image },
    });
    return data.updateQuestion;
  },

  deleteQuestion: async (id: string) => {
    const { data } = await apolloClient.mutate({
      mutation: gql`
        mutation DeleteQuestion($id: ID!) {
          deleteQuestion(id: $id)
        }
      `,
      variables: { id },
    });
    return data.deleteQuestion;
  },
};

File: flutter/lib/data/services/question_service.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:graphql_flutter/graphql_flutter.dart';
import '../lib/models/question_model.dart';
import 'auth_service.dart';

class QuestionService {
  final GraphQLClient _client = AuthService().client;
  final secureStorage = const FlutterSecureStorage();

  Future<Context?> _getHeadersContext() async {
    final String? token = await secureStorage.read(key: "jwt");

    if (token != null && token.isNotEmpty) {
      return Context.fromList([
        HttpLinkHeaders(headers: {'Authorization': 'Bearer $token'}),
      ]);
    }
    return null;
  }

  Future<List<Question>> getQuestions({required String quizId}) async {
    final String query = """
      query GetQuestions(\$quizId: ID!) {
        questions(quizId: \$quizId) {
          id
          quizId
          question
          image
          createdAt
        }
      }
    """;

    final QueryOptions options = QueryOptions(
      document: gql(query),
      variables: {'quizId': quizId},
      fetchPolicy: FetchPolicy.networkOnly,
      context: await _getHeadersContext(),
    );

    final QueryResult result = await _client.query(options);

    if (result.hasException) {
      print(result.exception.toString());
      throw Exception('Failed to load questions');
    }

    final List<dynamic> questionsJson = result.data?['questions'] ?? [];
    return questionsJson.map((json) => Question.fromJson(json)).toList();
  }
}
